## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [64]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd


### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [65]:
# Write your code for importing the csv file here
movies = pd.read_csv("Movies.csv")
movies.head(2)

color   director_name  num_critic_for_reviews  duration   
0  Color   James Cameron                   723.0     178.0  \
1  Color  Gore Verbinski                   302.0     169.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name   
0                      0.0                   855.0  Joel David Moore  \
1                    563.0                  1000.0     Orlando Bloom   

   actor_1_facebook_likes        gross                           genres  ...   
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...  \
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   

  num_user_for_reviews language  country  content_rating       budget   
0               3054.0  English      USA           PG-13  237000000.0  \
1               1238.0  English      USA           PG-13  300000000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio   
0      2009.0                  936.0        7.9          1.78  \
1      2007.0                 5000.0        7.1          2.35   

  movie_facebook_likes  
0                33000  
1                    0  

[2 rows x 28 columns]

In [66]:
movies.tail(2)

color director_name  num_critic_for_reviews  duration   
3851  Color  Edward Burns                    14.0      95.0  \
3852  Color      Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name   
3851                      0.0                   133.0  Caitlin FitzGerald  \
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes    gross        genres  ... num_user_for_reviews   
3851                   296.0   4584.0  Comedy|Drama  ...                 14.0  \
3852                    86.0  85222.0   Documentary  ...                 84.0   

     language  country  content_rating  budget  title_year   
3851  English      USA       Not Rated  9000.0      2011.0  \
3852  English      USA              PG  1100.0      2004.0   

     actor_2_facebook_likes imdb_score  aspect_ratio movie_facebook_likes  
3851                  205.0        6.4           NaN                  413  
3852                   23.0        6.6          1.85                  456  

[2 rows x 28 columns]

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [67]:
# Write your code for inspection here
movies.shape

(3853, 28)

In [68]:
movies.describe()

num_critic_for_reviews     duration  director_facebook_likes   
count             3852.000000  3852.000000              3853.000000  \
mean               163.046989   109.921080               784.507397   
std                123.927258    22.741728              3027.395958   
min                  1.000000    34.000000                 0.000000   
25%                 72.000000    95.000000                10.000000   
50%                134.000000   106.000000                58.000000   
75%                221.000000   120.000000               222.000000   
max                813.000000   330.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross   
count             3847.000000             3853.000000  3.853000e+03  \
mean               747.290096             7575.558526  5.096547e+07   
std               1841.623945            15405.383114  6.932033e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                183.000000              721.000000  6.830957e+06   
50%                427.000000             1000.000000  2.790000e+07   
75%                685.000000            12000.000000  6.550000e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster   
count     3.853000e+03                3853.000000           3847.000000  \
mean      1.024240e+05               11239.480145              1.378217   
std       1.502942e+05               18922.948757              2.056071   
min       5.000000e+00                   0.000000              0.000000   
25%       1.730900e+04                1817.000000              0.000000   
50%       5.052300e+04                3876.000000              1.000000   
75%       1.241850e+05               15972.000000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year   
count           3853.000000  3.853000e+03  3853.000000  \
mean             326.720478  4.524817e+07  2003.061511   
std              407.920904  2.234208e+08    10.007168   
min                1.000000  2.180000e+02  1920.000000   
25%              102.000000  1.000000e+07  1999.000000   
50%              203.000000  2.400000e+07  2005.000000   
75%              391.000000  5.000000e+07  2010.000000   
max             5060.000000  1.221550e+10  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             3852.000000  3853.000000   3781.000000           3853.000000  
mean              1959.018432     6.462886      2.109569           9089.546068  
std               4472.171290     1.053843      0.353236          21277.375223  
min                  0.000000     1.600000      1.180000              0.000000  
25%                362.000000     5.900000      1.850000              0.000000  
50%                664.000000     6.600000      2.350000            206.000000  
75%                971.000000     7.200000      2.350000          11000.000000  
max             137000.000000     9.300000     16.000000         349000.000000

In [69]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3851 non-null   object 
 1   director_name              3853 non-null   object 
 2   num_critic_for_reviews     3852 non-null   float64
 3   duration                   3852 non-null   float64
 4   director_facebook_likes    3853 non-null   float64
 5   actor_3_facebook_likes     3847 non-null   float64
 6   actor_2_name               3852 non-null   object 
 7   actor_1_facebook_likes     3853 non-null   float64
 8   gross                      3853 non-null   float64
 9   genres                     3853 non-null   object 
 10  actor_1_name               3853 non-null   object 
 11  movie_title                3853 non-null   object 
 12  num_voted_users            3853 non-null   int64  
 13  cast_total_facebook_likes  3853 non-null   int64

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

In [70]:
(movies.isnull().sum() > 0).sum()

12

### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [71]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations

movies.drop(["color",
"director_facebook_likes",
"actor_1_facebook_likes",
"actor_2_facebook_likes",
"actor_3_facebook_likes",
"actor_2_name",
"cast_total_facebook_likes",
"actor_3_name",
"duration",
"facenumber_in_poster",
"content_rating",
"country",
"movie_imdb_link",
"aspect_ratio",
"plot_keywords"], axis=1,inplace=True)
movies.head(2)

director_name  num_critic_for_reviews        gross   
0   James Cameron                   723.0  760505847.0  \
1  Gore Verbinski                   302.0  309404152.0   

                            genres actor_1_name   
0  Action|Adventure|Fantasy|Sci-Fi  CCH Pounder  \
1         Action|Adventure|Fantasy  Johnny Depp   

                                 movie_title  num_voted_users   
0                                    Avatar            886204  \
1  Pirates of the Caribbean: At World's End            471220   

   num_user_for_reviews language       budget  title_year  imdb_score   
0                3054.0  English  237000000.0      2009.0         7.9  \
1                1238.0  English  300000000.0      2007.0         7.1   

   movie_facebook_likes  
0                 33000  
1                     0

In [72]:
movies.columns.size

13

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'.

In [73]:
movies.isnull().sum()

director_name             0
num_critic_for_reviews    1
gross                     0
genres                    0
actor_1_name              0
movie_title               0
num_voted_users           0
num_user_for_reviews      0
language                  4
budget                    0
title_year                0
imdb_score                0
movie_facebook_likes      0
dtype: int64

In [74]:
len(movies)

3853

In [75]:
# Write you code here
movies.isnull().sum() * 100 / len(movies)

director_name             0.000000
num_critic_for_reviews    0.025954
gross                     0.000000
genres                    0.000000
actor_1_name              0.000000
movie_title               0.000000
num_voted_users           0.000000
num_user_for_reviews      0.000000
language                  0.103815
budget                    0.000000
title_year                0.000000
imdb_score                0.000000
movie_facebook_likes      0.000000
dtype: float64

In [76]:
movies["language"].value_counts()

language
English       3671
French          37
Spanish         26
Mandarin        14
German          13
Japanese        12
Hindi           10
Cantonese        8
Italian          7
Portuguese       5
Korean           5
Norwegian        4
Thai             3
Persian          3
Danish           3
Dutch            3
Dari             2
Indonesian       2
Hebrew           2
Aboriginal       2
Arabic           1
Russian          1
Vietnamese       1
Dzongkha         1
Romanian         1
Zulu             1
Bosnian          1
Czech            1
Icelandic        1
Hungarian        1
Mongolian        1
Aramaic          1
Telugu           1
Kazakh           1
Maya             1
Filipino         1
Swedish          1
Name: count, dtype: int64

#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [77]:
# Write your code for filling the NaN values in the 'language' column here
movies['language'].isnull().sum()

4

In [78]:
movies['language'].fillna('Engliish', inplace=True)
movies['language'].isnull().sum()

0

In [82]:
(movies['language'] == "English").sum()

3671

In [79]:
movies['language'].value_counts()

language
English       3671
French          37
Spanish         26
Mandarin        14
German          13
Japanese        12
Hindi           10
Cantonese        8
Italian          7
Portuguese       5
Korean           5
Engliish         4
Norwegian        4
Danish           3
Dutch            3
Persian          3
Thai             3
Dari             2
Indonesian       2
Hebrew           2
Aboriginal       2
Russian          1
Romanian         1
Vietnamese       1
Arabic           1
Dzongkha         1
Zulu             1
Aramaic          1
Czech            1
Telugu           1
Icelandic        1
Filipino         1
Hungarian        1
Maya             1
Bosnian          1
Mongolian        1
Kazakh           1
Swedish          1
Name: count, dtype: int64

In [ ]:
(movies['language'] == 'English').sum()

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [ ]:
# Write your code for unit conversion here


**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [ ]:
# Write your code for creating the profit column here


In [ ]:
# Write your code for sorting the dataframe here


In [ ]:
top10 = # Write your code to get the top 10 profiting movies here

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000.

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [ ]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe
# and name that dataframe as 'IMDb_Top_250'


#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [ ]:
# Write your code for creating three new dataframes here
Meryl_Streep = # Include all movies in which Meryl_Streep is the lead

In [ ]:
Leo_Caprio = # Include all movies in which Leo_Caprio is the lead

In [ ]:
Brad_Pitt = # Include all movies in which Brad_Pitt is the lead

In [ ]:
# Write your code for combining the three dataframes here
Combined =

In [ ]:
# Write your code for grouping the combined dataframe here


In [ ]:
# Write the code for finding the mean of critic reviews and audience reviews here


#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

##  Task2 Amazon Prime video data analysis


https://www.kaggle.com/datasets/shivamb/amazon-prime-movies-and-tv-shows?resource=download


Show uniques values of a column 'director'

show all unique values with their counts

get total no of uniwue values of whole data frame

In which year highest no of TV shows and movies were released

how many TV and Movie shows are there in Data frame

show all records with type 'movies; and country united kingdom

show all movie records directed by Paul

Show top 3 Directors, who gave highest no of TV shows and movies released on Prime video

In which year Highest rating show was there

### Task 3 Netflix Analysis
Information about TV shows and Movies
1- upload csv


2- describe, info,dtypes


3- uniques values of each column


4- total no of unique values of Dataframe


5- Unique values with their count


6-is any missing value with count

7- who is the director and show id of show #"ZOO"

8- Convert Datatype of column release date to DateTime

9-In which year highest no of TV shows and Movies relaesed

10-How many movies and TV shows are there in data set

11- Display Titles of all TV shows that were released in " United Sates" only

12- show top 10 Directors who gave highest no of TV shows and Movies on Netflix

13- show the record of all 'Horror' type of Movies

14 What are different 'Ratings' given by Netflix

15- What is Maximum duration of TV show on Netflix

16-sort dataframe by year

